src/test/hosting/config.spec.ts

```diff
@@ -0,0 +1,344 @@
+        want: [{ site: "site" }],
+      },
+      {
+        desc: "a hosting config with a non-hosting only flag",
+        cfg: [{ site: "site" }],
+        only: "functions",
+        want: [],
+      },
+    ];
+
+    for (const t of tests) {
+      it(`should be able to parse ${t.desc}`, () => {
+        if (t.wantErr) {
+          expect(() => config.filterOnly(t.cfg, t.only)).to.throw(FirebaseError, t.wantErr);
+        } else {
+          const got = config.filterOnly(t.cfg, t.only);
+          expect(got).to.deep.equal(t.want);
+        }
+      });
+    }
+  });
+
+  describe("with an except parameter, resolving targets", () => {
+    const tests: Array<
+      {
+        desc: string;
+        cfg: HostingMultiple;
+        except: string | undefined;
+      } & RequireAtLeastOne<{
+        want: HostingMultiple;
+        wantErr: RegExp;
+      }>
+    > = [
+      {
+        desc: "a hosting config with multiple sites, no targets, omitting the second site",
+        cfg: [{ site: "default-site" }, { site: "different-site" }],
+        except: `hosting:different-site`,
+        want: [{ site: "default-site" }],
+      },
+      {
+        desc: "a normal hosting config with a target, omitting the target",
+        cfg: [{ target: "main" }],
+        except: "hosting:main",
+        want: [],
+      },
+      {
+        desc: "a hosting config with multiple targets, omitting one",
+        cfg: [{ target: "t-one" }, { target: "t-two" }],
+        except: "hosting:t-two",
+        want: [{ target: "t-one" }],
+      },
+      {
+        desc: "a hosting config with multiple targets, omitting all hosting",
+        cfg: [{ target: "t-one" }, { target: "t-two" }],
+        except: "hosting",
+        want: [],
+      },
+      {
+        desc: "a hosting config with multiple targets, omitting an invalid target",
+        cfg: [{ target: "t-one" }, { target: "t-two" }],
+        except: "hosting:t-three",
+        want: [{ target: "t-one" }, { target: "t-two" }],
+      },
+      {
+        desc: "a hosting config with no excpet string",
+        cfg: [{ target: "target" }],
+        except: undefined,
+        want: [{ target: "target" }],
+      },
+      {
+        desc: "a hosting config with a non-hosting except string",
+        cfg: [{ target: "target" }],
+        except: "functions",
+        want: [{ target: "target" }],
+      },
+    ];
+
+    for (const t of tests) {
+      it(`should be able to parse ${t.desc}`, () => {
+        if (t.wantErr) {
+          expect(() => config.filterExcept(t.cfg, t.except)).to.throw(FirebaseError, t.wantErr);
+        } else {
+          const got = config.filterExcept(t.cfg, t.except);
+          expect(got).to.deep.equal(t.want);
+        }
+      });
+    }
+  });
+
+  const PUBLIC_DIR_ERROR_PREFIX = /Must supply a "public" directory/;
+  describe("validate", () => {
+    const tests: Array<{
+      desc: string;
+      site: HostingSingle;
+      wantErr: RegExp | false;
```

Comment:
idea: `wantErr?: RegExp` could eliminate all the places where you have `wantErr: false`. Follows a nice go-pattern of "if the default value works, don't bother specifying it". The check in the test would still be `!t.wantErr`, but you wouldn't have to specify `wantErr: false` in each happy test case (which should be pretty much assumed unless you have a desired error specified)

Reply:
Removed. I had done that to improve the readability of the test matrix. There is no desired outcome; there was no `want` for me to set when I expected things to go smoothly. So instead saying `wantErr: false` felt like I was making this test actually _assert_ something, as tests usually do.